java/WordCountOptimized.java

```diff
@@ -217,33 +207,66 @@ public static int tokenizeAndSubmitBlock(TObjectIntCustomHashMap<byte[]> counts,
         }
         return 0;
     }
-    
-    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
-        m.adjustOrPutValue(word, 1, 1);        
+
+    private static final int TABLE_SIZE = 1024 * 1024;
+    private static final int TABLE_MASK = TABLE_SIZE - 1;
+
+    static class WordMap {
+        final byte[][] keys = new byte[TABLE_SIZE][];
+        final int[] values = new int[TABLE_SIZE];
+        int size = 0;
     }
-    
+
+    private static int submitWord(byte[][] keys, int[] values, int size,
+                                   byte[] word, int off, int len) {
+        int hashCode = (int) XX_HASH.hashBytes(word, off, len);
+        int index = hashCode & TABLE_MASK;
```

Comment:
Ah, you're just doing linear probing in increasing order.  This is unsafe because you may run off the end of the bucket array.